File: 2. 两数相加/main.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  let head = new ListNode("head");
  let n = 0;
  let tmp = head;
  let sum = 0;

  while (l1 || l2) {
    let val1 = l1 ? l1.val : 0;
    let val2 = l2 ? l2.val : 0;

    sum = val1 + val2 + n;

    tmp.next = new ListNode(sum % 10); // 注意这里是对10 求余数
    tmp = tmp.next;

    n = sum >= 10 ? 1 : 0;

    if (l1) l1 = l1.next;
    if (l2) l2 = l2.next;
  }

  if (n > 0) tmp.next = new ListNode(n);
  return head.next;
};

File: 2. 两数相加/main.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val,
        }
    }
}
impl Solution {
    pub fn add_two_numbers(
        l1: Option<Box<ListNode>>,
        l2: Option<Box<ListNode>>
    ) -> Option<Box<ListNode>> {
        let mut l1 = l1; // ← 影子绑定为可变
        let mut l2 = l2; // ← 影子绑定为可变
        let mut carry = 0;
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy; // 指向结果链表尾部

        while l1.is_some() || l2.is_some() || carry != 0 {
            let mut sum = carry;

            if let Some(node) = l1 {
                sum += node.val;
                l1 = node.next; // 前进
            }

            if let Some(node) = l2 {
                sum += node.val;
                l2 = node.next; // 前进
            }

            carry = sum / 10;
            let digit = sum % 10;

            // 追加新节点
            tail.next = Some(Box::new(ListNode::new(digit)));
            // 安全地把 tail 移到新尾结点
            tail = tail.next.as_mut().unwrap();
        }

        dummy.next
    }
}
